akka-cluster-tools/src/main/scala/akka/cluster/singleton/ClusterSingletonManager.scala

```diff
@@ -422,6 +452,16 @@ class ClusterSingletonManager(
     removed = removed filter { case (_, deadline) ⇒ deadline.hasTimeLeft }
   }
 
+  // for CoordinatedShutdown
+  val coordShutdown = CoordinatedShutdown(context.system)
+  val memberExitingProgress = Promise[Done]()
+  coordShutdown.addTask(CoordinatedShutdown.PhaseClusterExiting)(() ⇒
+    memberExitingProgress.future)
+  coordShutdown.addTask(CoordinatedShutdown.PhaseClusterExiting) { () ⇒
+    implicit val timeout = Timeout(coordShutdown.timeout(CoordinatedShutdown.PhaseClusterExiting))
+    self.ask(SelfExiting).mapTo[Done]
+  }
```

Comment:
that is `preStart` of another actor :)

Yeah, that is a delicate difference. The `MemberExited` event is published immediately (leader may have performed that transition on other node), and that will trigger `run` of `CoordinatedShutdown`, while `PhaseClusterExiting` will happen later.  The concrete reason why I use `PhaseClusterExiting` in the singleton is that the graceful shutdown of sharding should preferably complete before stopping the singleton shard coordinator.